Step2 Restore Conflict

Hyeongyu-IM opened the pull request:
- 한방커밋 안하기...
- 좀더 개선을 어떻게 하면 좋을지... 고민

리뷰어님께
 - 제가 스태시 쓰는과정에서 오류가나서 레포를 다시 클론후에 세팅하였습니다 ㅜㅜ 
 - 스텝2에서 새로 생성된파일은 
    - StringExtension
    - StringOperation
    - StringCalcultorTest 입니다

Hyeongyu-IM committed:
Step2 Restore Conflict

- 한방커밋 안하기...
- 좀더 개선을 어떻게 하면 좋을지... 고민

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+
+import Foundation
+
+enum CalculatorError: Error {
```
Calculator 에러 타입을 구현하셨네요. 좋습니다! 👍

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+enum OperatorType: String {
```
Operator 타입을 구현하셨네요. 좋습니다! 👍

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+
+struct StringCalculator {
+    func startStringCalculate(data: String) -> Int  {
+        let splitString = data.split(separator: " ")
+                              .map { "\($0.trimmingCharacters(in: .whitespaces))" }
```
문자열을 공백 단위로 분리하고 공백을 제거하는 map 연산을 수행하셨네요. 혹시 한 번 더 공백을 제거하는 연산을 수행하신 이유가 있을까요? 의견이 궁금합니다.😁

Hyeongyu-IM replied:
아 이부분은 제가 메서드 이해를 잘못했네요 수정하겠습니다 ㅎㅎ 감사합니다

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+                                return false
+                            }
+        
+        return operators.count == isHasOperator.count ? true : false
```
시간이 지나서 다시 코드를 볼 때 `operators.count == isHasOperator.count` 조건을 보고 올바른 연산자를 의미하는지 이해하기 어려울 것 같습니다.
올바른 연산자인지 판단할 수 있도록 하는 더 좋은 표현법이 있을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+}
+
+struct StringCalculator {
+    func startStringCalculate(data: String) -> Int  {
```
다른 중괄호'{'와 다르게 띄어쓰기가 2번 작성되어 있는 이유가 있을까요?😲

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        fatalError(CalculatorError.error.description)
+    }
+
+    func isHasValidationOperator(data: [String]) -> Bool {
+        let operators = data.enumerated().filter { index, data in
+            if index % 2 == 1 {
+                return true
+            }
+            return false
+        }
+        let isHasOperator = operators
+                            .map { "\($0.element)" }
+                            .map { OperatorType(rawValue: $0) ?? .none }
+                            .filter {
+                                if $0 == .add || $0 == .minus || $0 == .multiple || $0 == .divide {
+                                    return true
+                                }
+                                return false
+                            }
```
메서드명을 보고 유추하면 올바른 연산자를 가지고 있는지 검증하는 메서드로 보입니다. 그런데 실제로는 메서드 내에서

1. 수식에서 홀수 번째 요소를 추출
2. 추출한 요소에서 사칙연산 연산자만을 다시 추출

위와 같이 다양한 일을 수행하는 것으로 보입니다.
메서드의 할일을 분리해보면서, 간결하게 표현해보는 방식으로 개선하려면 어떻게 하면 좋을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        return operators.count == isHasOperator.count ? true : false
+    }
+
+    func isHasIntAbleString(data: [String]) -> Bool {
+        let intStrings = data.enumerated().filter { index, data in
+            if index % 2 == 0 {
+                return true
+            }
+            return false
+        }
+        let isHasIntAbleString = intStrings
+            .map { "\($0.element)" }
+            .map { OperatorType(rawValue: $0) ?? .none }
+            .filter {
+                if $0 == .add || $0 == .minus || $0 == .multiple || $0 == .divide {
+                    return false
+                }
+                return true
+            }
```
메서드명을 보고 어떤 역할을 수행하는지 불명확해 보였습니다. 메서드의 역할을 조금 더 명확하게 표현할 수 있는 이름이 있을까요?🤔
그리고 메서드 내에서 많은 일을 수행하고 있는 것으로 보입니다.

1. 수식에서 짝수 번째 요소를 추출
2. 추출한 요소에서 사칙연산 연산자가 아닌 요소만 다시 추출

위와 같이 다양한 일을 수행하는 것으로 보입니다.
메서드의 할 일을 분리해보면서, 간결하게 표현해보는 방식으로 개선하려면 어떻게 하면 좋을까요?🤔

Hyeongyu-IM replied:
제가 PR을 처음해봐서 숨겨진걸 늦게 펼쳐보았네요 ㅠㅠ 

메서드는 분리해보았는데 이름을 어떻게 간결하게....?해야할까요...

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+                }
+                return true
+            }
+        return intStrings.count == isHasIntAbleString.count ? true : false
```
시간이 지나서 다시 코드를 볼 때 `intStrings.count == isHasIntAbleString.count` 조건을 보고 올바른 피연산자를 의미하는지 이해하기 어려울 것 같습니다.
올바른 피연산자인지 판단할 수 있도록 하는 더 좋은 표현법이 있을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        return intStrings.count == isHasIntAbleString.count ? true : false
+    }
+
+    func calculateString(data: [String]) -> Int {
```
`calculateString` 메서드 내에서도 많은 일을 수행하고 있는 것으로 보입니다.
역할을 분담할 수 있게 메서드를 분리해보는 것은 어떨까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        var lhs: Int?
+        var rhs: Int?
+        var operatorSymbol: OperatorType?
+        for i in data {
```
for 문 안에 `i` 라는 이름은 실제 코드상에서 어떤 역할을 하는지 알기 어려워 보입니다. 역할을 표현할 수 있는 더 좋은 이름이 있을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        var rhs: Int?
+        var operatorSymbol: OperatorType?
+        for i in data {
+            if i.toInt() == nil {
```
`i.toInt() == nil` 조건을 보고 연산자를 의미 또는 구분하는지 이해하기 어려울 것 같습니다. 조건을 보고 연산자라고 판단할 수 있는 더 좋은 표현법이 있을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+                guard let i = i.toInt() else { fatalError(CalculatorError.convertError.description) }
+                rhs = i
+            }
+            if lhs != nil && operatorSymbol != nil && rhs != nil {
```
`lhs != nil && operatorSymbol != nil && rhs != nil` 조건이 연산을 수행할 수 있다는 의미를 가지고 있는지 판별하기가 쉽지 않았습니다. 연산을 수행할 수 있는 상태를 나타낼 수 있도록 하는 조건이나 더 나은 표현법이 있을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+                rhs = i
+            }
+            if lhs != nil && operatorSymbol != nil && rhs != nil {
+                switch operatorSymbol {
+                case .add:
+                    lhs = add(lhs ?? 0, rhs ?? 0)
+                case .minus:
+                    lhs = minus(lhs ?? 0, rhs ?? 0)
+                case .multiple:
+                    lhs = multiple(lhs ?? 0, rhs ?? 0)
+                case .divide:
+                    lhs = divide(lhs ?? 0, rhs ?? 0)
+                
+                default:
+                    continue
+                }
+                rhs = nil
+            }
+        }
+        return lhs ?? 0
+    }
```
변수 `lhs`의 역할은 무엇일까요?🧐 피연산자를 의미하는 `lhs` 에 연산의 결과가 계속 반영되고 수정되고 있습니다. `lhs`의 역할이 한 가지만 수행할 수 있도록 하면 `lhs` 역할이 명확해지지 않을까 생각합니다.😊 역할을 한가지만 수행할 수 있도록 하려면 구조를 어떻게 변경하면 좋을까요?🤔

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        return lhs + rhs
+    }
+
+    func minus(_ lhs: Int,_ rhs: Int) -> Int {
+        return lhs - rhs
+    }
```
`minus` 라는 이름은 빼기 부호를 나타내는 명사의 느낌이 조금 더 강하게 들었습니다. 빼기를 수행한다는 의미를 나타낼 수 있는 메서드명으로 바꿔보는 것은 어떨까요?😁

zdodev commented on RacingCar/Step2_StringOperation.swift:
```diff
@@ -0,0 +1,150 @@
+        return lhs - rhs
+    }
+
+    func multiple(_ lhs: Int,_ rhs: Int) -> Int {
+        return lhs * rhs
+    }
```
`multiple` 이라는 이름은 배수를 나타내는 명사의 느낌이 들었습니다. 곱셈을 수행한다는 의미를 나타낼 수 있는 메서드명으로 바꿔보는 것은 어떨까요?😁

zdodev commented on RacingCar/StringExtension.swift:
```diff
@@ -0,0 +1,21 @@
+
+extension String {
+    func toInt() -> Int? {
+        if let toint = Int(self) {
```
옵셔널 바인딩된 상수명이 `toint` 입니다! 혹시 의도적으로 int로 작성하셨는지 의견이 궁금합니다!😲

zdodev commented on RacingCar/StringExtension.swift:
```diff
@@ -0,0 +1,21 @@
+        return nil
+    }
+    
+    func isNotEmpty() -> Bool {
+        return self.isEmpty ? false : true
+    }
```
`isNotEmpty()` 메서드를 추가로 작성해주셨네요! `isEmpty` 메서드를 사용하는 대신 extension을 통해 `isNotEmpty()` 메서드를 작성하신 이유에 대한 의견을 들어보고 싶습니다!😁

Hyeongyu-IM replied:
이 부분은 제가 주로 데이터가 비었는지 확인할때 guard문을 활용하는데 isEmpty를 쓰면 앞에 !를 붙여야해서 가독성이 떨어진다 생각하여 추가하였습니다 ㅎㅎ

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,49 @@
+@testable import RacingCar
+
+class StringCalculatorTest: XCTestCase {
+    let cal = StringCalculator()
```
`cal`이라는 상수명은 어떤 의미를 담고 있는 이름일까요?🤔 조금 더 명확하게 이해할 수 있는 이름을 지어보는 것은 어떨까요?😁

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,49 @@
+class StringCalculatorTest: XCTestCase {
+    let cal = StringCalculator()
+
+    func test_isHasOperator() {
```
문자열 배열에 연산자 존재 여부에 대한 테스트를 작성해주셨네요. 좋습니다!👍

테스트의 이름에서 `is`를 제외하고 연산자의 존재 여부를 판단할 수 있는 이름인 `hasOperator` 또는 `has_operator` 도 좋다는 생각이 들었습니다. 이 부분은 `isHas` 로 작성하신 메서드 리뷰의 공통사항입니다~
이 부분에 대해서 @Hyeongyu-IM 님의 의견이 궁금합니다. @Hyeongyu-IM의 생각에 따라 `isHas`로 작성하고 진행하셔도 무방합니다.😄

Hyeongyu-IM replied:
값을 판별하는 부분에는 항상 is를 붙여왔는데 테스트 케이스는 직접작성한게 처음이라 어떻게 해야할지 잘모르겠네요 ㅎㅎ 의견 감사합니다

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,49 @@
+        XCTAssertEqual(cal.isHasIntAbleString(data: ["1", "5", "2", "1"]), true)
+    }
+    
+    func test_Add() {
```
테스트의 메서드명이 대문자 `Add`로 시작하네요. 위의 테스트 네이밍과 컨벤션을 맞춰볼 수 있을까요?
그리고 단순 동사보다는 어떤 테스트를 수행하는지에 대한 간략한 내용이 포함되어 있으면 더욱 테스트 가독성이 좋겠다는 생각이 들었습니다.😁

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,49 @@
+        XCTAssertEqual(cal.add(5, 50), 55)
+    }
+    
+    func test_Minus() {
```
테스트의 이름이 대문자 `Minus`로 시작하네요. 위의 테스트 네이밍과 컨벤션을 맞춰볼 수 있을까요?
어떤 테스트를 수행하는지에 대한 간략한 내용이 포함되어 있으면 더욱 테스트 가독성이 좋겠다는 생각이 들었습니다.😁

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,49 @@
+        XCTAssertEqual(cal.minus(5, 10), -5)
+    }
+    
+    func test_Multiple() {
```
테스트의 이름이 대문자 `Multiple`로 시작하네요. 위의 테스트 네이밍과 컨벤션을 맞춰볼 수 있을까요?
어떤 테스트를 수행하는지에 대한 간략한 내용이 포함되어 있으면 더욱 테스트 가독성이 좋겠다는 생각이 들었습니다.😁

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,49 @@
+        XCTAssertEqual(cal.multiple(5, 50), 250)
+    }
+    
+    func test_Divide() {
```
테스트의 이름이 대문자 `Divide`로 시작하네요. 위의 테스트 네이밍과 컨벤션을 맞춰볼 수 있을까요?
어떤 테스트를 수행하는지에 대한 간략한 내용이 포함되어 있으면 더욱 테스트 가독성이 좋겠다는 생각이 들었습니다.😁

zdodev reviewed (changes requested):
@Hyeongyu-IM 안녕하세요~

본격적인 과정시작의 첫 PR을 장식하셨네요~🤗
앞으로도 과감하게 PR 보내주세요. ~

step2 브랜치에서 본인의 메인 브랜치로 PR 잘보내주셨습니다.

사용자 입력에 대한 처리도 신경써서 해주신 점과 
전체적으로 else 구문을 사용하지 않고 작성하려고 하신 점이 눈에 들어왔습니다.👍

테스트를 잘 작성해주셨는데 추가적으로 사용자가 입력하는 수식에 대한 테스트도 작성되어 있으면 더욱 좋겠다는 생각이 들었습니다.

그 이외에 몇 가지 궁금증과 질문에 대해 의견 남겨보았습니다~
이해가 잘 안되거나 어려운 점에 대해서 공유해주시면 이야기 나눠주시면 좋겠습니다.😊

Hyeongyu-IM committed:
Fix - 동일한 공백처리 연산자 수정

Hyeongyu-IM committed:
Fix - 이후에 코드를 볼때 잘 알아볼수있도록 네이밍 수정

Hyeongyu-IM committed:
Fix - 두번연속 띄어진 부분 수정

Hyeongyu-IM committed:
Fix - 테스트 케이스 대문자 시작부분 수정

Hyeongyu-IM commented:
수정을 마친후에 제가 다시 한번 코멘트를 읽어보았는데요 

" 추가적으로 사용자가 입력하는 수식에 대한 테스트도 작성되어 있으면 더욱 좋겠다는 생각이 들었습니다 "

이부분이 잘 이해가 안되서요 혹시 readLine에 대한 테스트를 말씀하시는 걸까요? 

수식에 대한 테스트라면 어떤 부분이 될수있을까요? 현재는 숫자와 알맞은 자리에 수식을 입력했는지 확인하고 있는데 어떻게 개선할수 있을까요??


------------------------

전체적으로 코드를 조금 개선해 보았습니다 

그런데 마지막 테스트 케이스를 작성하는 과정에서 해당 객체의 userInput을 메서드가 파라미터를 받지않고 self를 통해 값을 연산했는데 

이럴경우 테스트 케이스 작성에 조금 에로사항이 생기더라구요

StringCalculator에서는 호출해야할 메서드 말고는 private시켜주고, 유효성검사는 하나하나 테스트 케이스를 작성했는데요

보통 테스트 케이스작성시에 이렇게 private는 어떻게 처리하는지? 그리고 메서드 input이 없는경우에는 테스트가 불편한데 이부분은 어떻게 해결하는지 궁금합니다

Hyeongyu-IM committed:
Fix - 메서드안 두개의 역할 분리

Hyeongyu-IM committed:
Refactoring

1. Step2를 폴더를 나누었습니다
2. 유효성검사를 객체를 나누어 분리하였습니다
3. 유효성 프로토콜을 생성했습니다
4. 계산기의 실질적인 수식을 단순화 했습니다

Hyeongyu-IM committed:
Fix

StringCalculator 로직부분 오류 수정

Hyeongyu-IM committed:
testCase

StringValidationChecker TestCase작성

Hyeongyu-IM committed:
testCase

StringCalculator testCase

zdodev commented:
수정을 마친후에 제가 다시 한번 코멘트를 읽어보았는데요.
" 추가적으로 사용자가 입력하는 수식에 대한 테스트도 작성되어 있으면 더욱 좋겠다는 생각이 들었습니다 "
이부분이 잘 이해가 안되서요 혹시 readLine에 대한 테스트를 말씀하시는 걸까요?

이 부분은 배열 형태가 아닌 수식 문자열 그대로("2 + 3 - 4 * 5" 등)를 입력하고 테스트 해보는 것에 대한 내용이었습니다~

그리고 추가로 수정해주신 부분 리뷰하고 답변드리겠습니다~

zdodev commented on RacingCar/Step_2/Step2_StringValidationChecker.swift:
```diff
@@ -0,0 +1,68 @@
+
+import Foundation
+
+struct StringCalculatorVaildationChecker: StringValidationChecker {
```
입력값 유효성 검사를 담당하는 타입을 구현하셨네요. 훨씬 보기 좋아졌습니다.👍

zdodev commented on RacingCar/Step_2/Step2_StringValidationChecker.swift:
```diff
@@ -0,0 +1,68 @@
+            }
+            return false
+        }.map { "\($0.element)" }
+        return operatorbleData
+    }
+
+    var intAbleData: [String] {
+        let splitStrings = inputString.split(separator: " ").map { "\($0)" }
+        let intAbleStrings = splitStrings.enumerated().filter { index, data in
+            if index % 2 == 0 {
+                return true
+            }
+            return false
+        }.map { "\($0.element)" }
+        return intAbleStrings
+    }
+    
+    func isHasValidationOperator() -> Bool {
+        let isOperator = self.operatorbleData
+            .map { OperatorType(rawValue: $0) ?? .none }
+            .filter {
+            if $0 == .add || $0 == .subtract || $0 == .multiply || $0 == .divide {
+                return true
+            }
+            return false
+        }
+        return self.operatorbleData.count == isOperator.count ? true : false
+    }
+
+    func isHasIntAbleString() -> Bool {
+        let isHasIntAbleString = self.intAbleData
+            .map { OperatorType(rawValue: $0) ?? .none }
+            .filter {
+            if $0 == .add || $0 == .subtract || $0 == .multiply || $0 == .divide {
+                return false
+            }
+            return true
+        }
+        return self.intAbleData.count == isHasIntAbleString.count ? true : false
+    }
```
보다 명확한 역할 분리와 네이밍으로 전보다 읽기 좋아졌습니다.👍

zdodev commented on RacingTest/StringCalculatorTest.swift:
```diff
@@ -0,0 +1,52 @@
+    
+    func test_add_input() {
+        let calculator = StringCalculator(userInput: "1 + 5")
+        XCTAssertEqual(calculator.startStringCalculate(), 6)
+    }
+    
+    func test_subtract_input() {
+        let calculator = StringCalculator(userInput: "5 - 1")
+        XCTAssertEqual(calculator.startStringCalculate(), 4)
+    }
+    
+    func test_multiply_input() {
+        let calculator = StringCalculator(userInput: "1 * 5")
+        XCTAssertEqual(calculator.startStringCalculate(), 5)
+    }
+    
+    func test_divide_input() {
+        let calculator = StringCalculator(userInput: "5 / 5")
+        XCTAssertEqual(calculator.startStringCalculate(), 1)
+    }
+    
+    func test_two_operator() {
+        let calculator = StringCalculator(userInput: "5 / 5 * 10")
+        XCTAssertEqual(calculator.startStringCalculate(), 10)
+    }
+    
+    func test_two_operator_2() {
+        let calculator = StringCalculator(userInput: "5 * 5 + 10")
+        XCTAssertEqual(calculator.startStringCalculate(), 35)
+    }
+    
+    func test_two_operator_3() {
+        let calculator = StringCalculator(userInput: "5 + 5 - 10")
+        XCTAssertEqual(calculator.startStringCalculate(), 0)
+    }
+    
+    func test_two_operator_4() {
+        let calculator = StringCalculator(userInput: "5 / 5 + 1")
+        XCTAssertEqual(calculator.startStringCalculate(), 2)
+    }
```
사용자 입력값에 대한 다양한 테스트 작성 좋습니다.👍 테스트 메서드마다 assert문을 하나씩 달아주어서 보기 편했습니다~

zdodev commented on RacingTest/StringValidationCheckerTest.swift:
```diff
@@ -0,0 +1,36 @@
+
+class StringValidationCheckerTest: XCTestCase {
+    
+    func test_noneHasOperator() {
+        let checker = StringCalculatorVaildationChecker(userInput: "1 2 3 4 5")
+        XCTAssertEqual(checker.intAbleData, ["1", "3", "5"])
+        XCTAssertEqual(checker.operatorbleData, ["2", "4"])
+        XCTAssertEqual(checker.isHasValidationOperator(), false)
+        XCTAssertEqual(checker.isHasIntAbleString(), true)
+    }
+    
+    func test_hasOperator() {
+        let checker = StringCalculatorVaildationChecker(userInput: "1 + 3 - 5")
+        XCTAssertEqual(checker.intAbleData, ["1", "3", "5"])
+        XCTAssertEqual(checker.operatorbleData, ["+", "-"])
+        XCTAssertEqual(checker.isHasValidationOperator(), true)
+        XCTAssertEqual(checker.isHasIntAbleString(), true)
+    }
+    
+    func test_longUserInput() {
+        let checker = StringCalculatorVaildationChecker(userInput: "1 * 4 / 4 * 10 + 1 + 2")
+        XCTAssertEqual(checker.intAbleData, ["1", "4", "4", "10", "1", "2"])
+        XCTAssertEqual(checker.operatorbleData, ["*", "/", "*", "+", "+"])
+        XCTAssertEqual(checker.isHasValidationOperator(), true)
+        XCTAssertEqual(checker.isHasIntAbleString(), true)
+    }
```
연산자와 피연산자 그리고 유효성 검사에 대한 테스트도 작성하셨네요. 좋습니다.👍👍

Hyeongyu-IM replied:
@zdodev 빠른답변 감사합니다 아직 해결되지 않은 궁금증이 있어서요

StringValidationChecker 테스트 케이스와 StringCalculatorTestCase를 보시면 아시겠지만 StringCalculator private를 이용하고 하나의 기능만 노출해서 테스트 케이스가 한개씩 작성되었지만 

StringValidationChecker는 internal로 선언되어 테스트 하기 용이했습니다 

또한 개선 전에는 검증과 연산을 담당하는 메서드에 파라미터를 넣어서 인풋과 아웃풋이 있어 테스트 하기 용이했지만 개선된 버전에서는 파라미터를 제거하고 struct전역변수를 이용해 연산했기 때문에 테스트 하기에는 조금 불편했습니다 

리뷰어님이 보시기에는 테스트를 위해서는 private를 internal로 노출시키는것과 메서드 파라미터를 전역으로 수행해 인풋이 없어 테스트가 어려운 메서드는 어떻게 생각하시는지 의견이 궁금합니다

zdodev replied:
private를 internal로 노출시키는 것과 파라미터를 전역으로 수행해 인풋이 없다는 것은 구조에 따라 다르지 않을까 생각합니다.

`private`은 외부에 노출시키지 않고 타입 내부에서 사용하기 위해 선언한다고 생각합니다. 그리고 보통 `private`끼리 쓰기보다는 `internal` 접근 레벨 이상의 메서드에서 사용하게 됩니다. 외부에서 타입의 `internal` 접근 레벨 이상의 메서드를 호출할 때 타입 내부적으로는 `private`의 요소에 접근 및 동작의 호출을 유발할 수 있고, 그 결과가 `internal` 접근 레벨 이상의 메서드의 반환값에 영향을 미칠 수 있습니다. 따라서 `private`을 테스트를 위해서 접근 레벨을 높이는 것보다는 그 동작의 호출을 유발시키는 `internal` 메서드 호출을 통해서 `private`의 테스트를 수행해 볼 수 있지 않을까 싶습니다. 모든 경우에 대해서 "`private`은 `internal`로 테스트할 수 있다." 라고 단정지을 수는 없지만, "테스트를 위해서 `private`의 접근 레벨을 높이고 싶은데?" 라는 생각이 든다면 한 번 더 타입과 메서드에 대한 구조를 생각해 볼 필요가 있지 않을까 싶습니다.

struct 전역 프로퍼티도 비슷하지 않을까 싶습니다. 인풋 파라미터가 없는 메서드 조차 어떠한 행위를 수행하고 결과를 반환하게 되는데 그런 메서드도 전역 프로퍼티를 사용해서 어떠한 행위를 수행할 수 있고 그 결과를 반환합니다. 인풋 파라미터가 없다는 것은 외부에서 어떠한 값이 필요없다는 것이고 내부 프로퍼티만을 사용해 반환값을 만들 수 있도록 구현한 사람이 의도하지 않았을까 싶습니다. 만약에 인풋 파라미터가 없어서 무언가 어려운 경우가 생겼다면 **메서드와 그 타입의 구조를 다시 한 번 둘러보고 다르게 바꿔보면 더 좋은 구조로 바꿀 수 없을까** 라고 생각해 볼 수 있는 계기가 될 수 있지 않을까 싶습니다.🙂

답변이 되었을까 싶습니다. 이게 정답이라고는 생각하지 않습니다. 이러한 고민을 한다는 게 좋은 현상이라고 생각합니다. 덕분에 저도 많은 고민을 해볼 수 있었습니다.😄 앞으로 진행하면서 함께 답을 찾아나가면 어떨까 싶습니다.👍

zdodev reviewed (approved):
전보다 네이밍과 역할에 대해서 명확하게 분리되어서 훨씬 보기 편했습니다
그리고 다양한 경우에 대한 추가적인 테스트 케이스 작성도 아주 좋았습니다.😊

이번 스텝은 여기서 마무리하고 다음 스텝 진행해주시면 될 것 같습니다! 고생하셨습니다!
